$OpenBSD: patch-common_parseconf_c,v 1.1 2012/06/01 07:12:42 jasper Exp $

Security fix for CVE-2012-2944: upsd can be remotely crashed
Patch taken from http://trac.networkupstools.org/projects/nut/changeset/3633

--- common/parseconf.c.orig	Fri Jun  1 09:03:34 2012
+++ common/parseconf.c	Fri Jun  1 09:04:40 2012
@@ -171,6 +171,13 @@ static void addchar(PCONF_CTX_t *ctx)
 
 	wbuflen = strlen(ctx->wordbuf);
 
+	/* CVE-2012-2944: only allow the subset Ascii charset from Space to ~ */
+	if ((ctx->ch < 0x20) || (ctx->ch > 0x7f)) {
+		fprintf(stderr, "addchar: discarding invalid character (0x%02x)!\n",
+			ctx->ch);
+		return;
+	}
+
 	if (ctx->wordlen_limit != 0) {
 		if (wbuflen >= ctx->wordlen_limit) {
 
